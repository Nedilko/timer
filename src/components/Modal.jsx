import Panel from './Panel'
import PropTypes from 'prop-types'

Modal.propTypes = {
  content: PropTypes.any
}

function Modal({content}) {
  return (
    <Panel>
      {content}
    </Panel>
  )
}

export default Modal
