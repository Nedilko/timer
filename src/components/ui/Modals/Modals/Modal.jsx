import { useEffect, useRef } from "react";
import { createPortal } from "react-dom";
import Panel from "../Panels/Panel";
import OverlayBackground from "../OverlayBackgrounds/OverlayBackground";
import PropTypes from "prop-types";
import Button from "../../Buttons/Button";

Modal.propTypes = {
  children: PropTypes.node.isRequired,
  title: PropTypes.string.isRequired,
  onApply: PropTypes.func.isRequired,
  onCancel: PropTypes.func.isRequired,
};

function Modal({ children, onApply, onCancel, title }) {
  const element = useRef(document.createElement("div"));

  useEffect(() => {
    const modalPortal = document.getElementById("modal-portal");
    const currentElement = element.current;

    modalPortal.appendChild(currentElement);

    return () => {
      modalPortal.removeChild(currentElement);
    };
  }, []);

  return createPortal(
    <OverlayBackground onClick={onCancel}>
      <Panel>
        <header role="modal" className="border-b border-gray-500 text-xl py-2">
          {title}
        </header>
        <main>{children}</main>
        <footer className="flex border-t border-gray-500 justify-end  py-2">
          <Button title={"Apply"} onClick={onApply} />
          <Button title={"Cancel"} onClick={onCancel} />
        </footer>
      </Panel>
    </OverlayBackground>,
    element.current
  );
}

export default Modal;
