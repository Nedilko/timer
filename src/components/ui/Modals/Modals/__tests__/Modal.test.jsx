import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import Modal from "../Modal";

describe("Modal", () => {
  const onApply = jest.fn();
  const onCancel = jest.fn();

  beforeAll(() => {
    const root = document.createElement("div");
    root.setAttribute("id", "modal-portal");
    document.body.appendChild(root);
  });

  beforeEach(() => {
    onCancel.mockClear();
    onApply.mockClear();
  });

  it("should match snapshot", () => {
    render(
      <Modal onApply={() => {}} onCancel={() => {}} title={"test title"}>
        <div>1</div>
        <div>2</div>
        <div>3</div>
      </Modal>
    );

    expect(screen.getByRole("modal")).toMatchSnapshot();
  });

  it("should call onApply", () => {
    render(
      <Modal onCancel={onCancel} onApply={onApply} title={"test title"}>
        <div>1</div>
        <div>2</div>
        <div>3</div>
      </Modal>
    );

    userEvent.click(screen.getByText(/Apply/i, { role: "button" }));

    expect(onApply).toBeCalled();
  });

  it("should call onCancel", () => {
    render(
      <Modal onCancel={onCancel} onApply={onApply} title={"test title"}>
        <div>1</div>
        <div>2</div>
        <div>3</div>
      </Modal>
    );

    userEvent.click(screen.getByText(/Cancel/i, { role: "button" }));

    expect(onCancel).toBeCalled();
  });

  it("shuld render title", () => {
    render(
      <Modal onCancel={onCancel} onApply={onApply} title={"test title"}>
        <div>1</div>
        <div>2</div>
        <div>3</div>
      </Modal>
    );

    expect(screen.getByRole("modal")).toHaveTextContent("test title");
  });
});
