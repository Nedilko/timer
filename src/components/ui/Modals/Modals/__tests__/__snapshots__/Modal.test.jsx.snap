// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal should match snapshot 1`] = `
<header
  class="border-b border-gray-500 text-xl py-2"
  role="modal"
>
  test title
</header>
`;
